src/_zkapauthorizer/sql.py

```diff
@@ -5,15 +5,98 @@
 to support testing the replication/recovery system.
 """
 
+from __future__ import annotations
+
 from enum import Enum, auto
-from sqlite3 import Cursor
-from typing import Union
+from sqlite3 import Connection as _SQLite3Connection
+from typing import Any, ContextManager, Iterable, Optional, Protocol, Union
 
-from attrs import define
+from attrs import frozen
 
 SQLType = Union[int, float, str, bytes, None]
 
 
+class AbstractCursor(Protocol):
```

Comment:
codify Cursor and Connection protocols here so we can frequently refer to the abstract protocol instead of a concrete type

Reply:
these are constrained by the actual `sqlite3.Connection` and `sqlite3.Cursor` types because we need them to unify with this protocol.